server/src/test/java/org/elasticsearch/common/util/CancellableSingleObjectCacheTests.java

```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.util;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.StepListener;
+import org.elasticsearch.action.support.PlainActionFuture;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.tasks.TaskCancelledException;
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.threadpool.TestThreadPool;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.LinkedList;
+import java.util.Objects;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.not;
+import static org.hamcrest.Matchers.sameInstance;
+
+public class CancellableSingleObjectCacheTests extends ESTestCase {
+
+    public void testNoPendingRefreshIfAlreadyCancelled() {
+        final TestCache testCache = new TestCache();
+        final PlainActionFuture<Integer> future = new PlainActionFuture<>();
+        testCache.get("foo", () -> true, future);
+        testCache.assertPendingRefreshes(0);
+        assertTrue(future.isDone());
+        expectThrows(ExecutionException.class, TaskCancelledException.class, future::get);
+    }
+
+    public void testListenersCompletedByRefresh() {
+        final TestCache testCache = new TestCache();
+
+        // The first get() calls the refresh function
+        final PlainActionFuture<Integer> future0 = new PlainActionFuture<>();
```

Comment:
Could be nice to use a subclass that throws if `done()` is called more than once, ensuring no double invoke on the listener, used throughout.

Reply:
Yep we can do that, see e09535e23b9f3d3d4f6bdb8a83774a7a7df42fca.